File: app/src/main/java/com/visiontutor/finalproject/SignUpFragment.java
package com.visiontutor.finalproject;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ToggleButton;

import com.androidnetworking.AndroidNetworking;
import com.androidnetworking.common.Priority;
import com.androidnetworking.error.ANError;
import com.androidnetworking.interfaces.JSONObjectRequestListener;
import com.androidnetworking.interfaces.ParsedRequestListener;
import com.visiontutor.finalproject.utils.City;
import com.visiontutor.finalproject.utils.URLS;
import com.visiontutor.finalproject.utils.Utils;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SignUpFragment extends Fragment implements OnClickListener {
    private View view;
    private EditText fullName, fullName2, emailId, mobileNumber,
            password, confirmPassword;
    private TextView login;
    private Button signUpButton;
    private CheckBox terms_conditions;
    private ToggleButton signuptoggle;
    private City _city;
    private Spinner citySpinner;
    private ArrayAdapter<City> cityArrayAdapter;

    public SignUpFragment() {

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.signup_layout, container, false);
        initViews();
        setListeners();
        AndroidNetworking.initialize(view.getContext());
        getCities();
        citySpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (position > 0) {
                    _city = (City) citySpinner.getSelectedItem();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {

            }

        });
        return view;
    }

    private void getCities() {

        citySpinner.setEnabled(false);

        AndroidNetworking.get(URLS.CITIES)
                .setPriority(Priority.HIGH)
                .build()
                .getAsObjectList(City.class, new ParsedRequestListener<List<City>>() {
                    @Override
                    public void onResponse(List<City> response) {
                        Log.d("SearchTutor.getCities", "onResponse: entered");
                        cityArrayAdapter = new ArrayAdapter<>(Objects.requireNonNull(getContext()),
                                android.R.layout.simple_spinner_dropdown_item,
                                response);
                        citySpinner.setAdapter(cityArrayAdapter);
                    }

                    @Override
                    public void onError(ANError anError) {
                        Log.d("SearchTutor.getCities", "onError: " + anError.getErrorDetail());
                    }
                });

        citySpinner.setEnabled(true);

    }

    // Initialize all views
    private void initViews() {
        fullName = view.findViewById(R.id.fullName);
        fullName2 = view.findViewById(R.id.fullName2);
        emailId = view.findViewById(R.id.userEmailId);
        mobileNumber = view.findViewById(R.id.mobileNumber);
        password = view.findViewById(R.id.password);
        confirmPassword = view.findViewById(R.id.confirmPassword);
        signUpButton = view.findViewById(R.id.signUpBtn);
        login = view.findViewById(R.id.already_user);
        terms_conditions = view.findViewById(R.id.terms_conditions);
        signuptoggle = view.findViewById(R.id.signuptoggle);
        citySpinner = view.findViewById(R.id.cities);
    }

    // Set Listeners
    private void setListeners() {
        signUpButton.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.signUpBtn:

                // Call checkValidation method
                checkValidation();
                break;

//            case R.id.already_user:
//
//                // Replace login fragment
//                new MainActivity().replaceLoginFragment();
//                break;
        }

    }

    // Check Validation Method
    private void checkValidation() {

        // Get all edittext texts
        String getFullName = fullName.getText().toString();
        String getFullName2 = fullName2.getText().toString();
        String getEmailId = emailId.getText().toString();
        String getMobileNumber = mobileNumber.getText().toString();
        int getCity = _city.getCityId();

        String getPassword = password.getText().toString();
        String getConfirmPassword = confirmPassword.getText().toString();
        boolean mode = signuptoggle.isChecked();
        // Pattern match for email id
        Pattern p = Pattern.compile(Utils.regEx);
        Matcher m = p.matcher(getEmailId);

        // Check if all strings are null or not
        if (getFullName.isEmpty()
                || getFullName2.isEmpty()
                || getEmailId.isEmpty()
                || getMobileNumber.isEmpty()
                || getCity <= 0
                || getPassword.isEmpty()
                || getConfirmPassword.isEmpty())

            new CustomToast().Show_Toast(getActivity(), view,
                    "All fields are required.");

            // Check if email id valid or not
        else if (!m.find())
            new CustomToast().Show_Toast(getActivity(), view,
                    "Your Email Id is Invalid.");

            // Check if both password should be equal
        else if (!getConfirmPassword.equals(getPassword))
            new CustomToast().Show_Toast(getActivity(), view,
                    "Both password doesn't match.");

            // Make sure user should check Terms and Conditions checkbox
        else if (!terms_conditions.isChecked())
            new CustomToast().Show_Toast(getActivity(), view,
                    "Please select Terms and Conditions.");

            // Else do signup or do your stuff
        else {
            if (mode)
                doStudentRegister(getCity,getFullName, getFullName2, getEmailId, getMobileNumber, getPassword);
            else
                doTutorRegister(getCity,getFullName, getFullName2, getEmailId, getMobileNumber, getPassword);
        }

    }

    private void doStudentRegister(int getCity, String getFullName, String getFullName2, String getEmailId, String getMobileNumber, String getPassword) {
        AndroidNetworking.post(URLS.STUDENT_REGISTER)
                .addBodyParameter("fname", getFullName)
                .addBodyParameter("lname", getFullName2)
                .addBodyParameter("email", getEmailId)
                .addBodyParameter("mobile", getMobileNumber)
                .addBodyParameter("city", String.valueOf(getCity))
                .addBodyParameter("password", getPassword)
                .build()
                .getAsJSONObject(new JSONObjectRequestListener() {
                    @Override
                    public void onResponse(JSONObject response) {
                        Log.d("SignUpFragment", "onResponse: entered");
                        try {
                            if (response.getString("status").equals("accepted")) {
                                Toast.makeText(getContext(),response.getString("message"),Toast.LENGTH_SHORT).show();
                                Intent intent = new Intent(getActivity(), StudentDashboard.class);
                                Log.d("StuSignUpFragment", "onResponse: accepted");
                                startActivity(intent);
                            } else {
                                new CustomToast().Show_Toast(getActivity(), view,
                                        response.getString("message"));
                            }
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }

                    @Override
                    public void onError(ANError anError) {
                        Log.d("StuSignUpFragment", "onError: " + anError.getErrorDetail());
                        Toast.makeText(getContext(), "error", Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void doTutorRegister(int getCity, String getFullName, String getFullName2, String getEmailId, String getMobileNumber, String getPassword) {

    }
}

File: app/src/main/java/com/visiontutor/app/SignUpFragment.kt
package com.visiontutor.app

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.transaction
import com.androidnetworking.AndroidNetworking
import com.androidnetworking.common.Priority
import com.androidnetworking.error.ANError
import com.androidnetworking.interfaces.JSONObjectRequestListener
import com.androidnetworking.interfaces.ParsedRequestListener
import com.visiontutor.app.utils.City
import com.visiontutor.app.utils.URLS
import com.visiontutor.app.utils.Utils
import kotlinx.android.synthetic.main.signup_layout.*
import org.json.JSONException
import org.json.JSONObject
import java.util.*
import java.util.regex.Pattern

class SignUpFragment : Fragment(), OnClickListener {

    private lateinit var baseContext:Context
    private lateinit var _city: City
    private lateinit var citySpinner: Spinner
    private lateinit var cityArrayAdapter: ArrayAdapter<City>

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        baseContext = requireContext()
        getCities()
        citySpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                if (position > 0) {
                    _city = citySpinner.selectedItem as City
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>) {

            }

        }
        return inflater.inflate(R.layout.signup_layout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        signup_btn.setOnClickListener(this)
    }

    private fun getCities() {

        citySpinner.isEnabled = false

        AndroidNetworking.get(URLS.CITIES)
                .setPriority(Priority.HIGH)
                .build()
                .getAsObjectList(City::class.java, object : ParsedRequestListener<List<City>> {
                    override fun onResponse(response: List<City>) {
                        Log.d("SearchTutor.getCities", "onResponse: entered")
                        cityArrayAdapter = ArrayAdapter(Objects.requireNonNull<Context>(context),
                                android.R.layout.simple_spinner_dropdown_item,
                                response)
                        citySpinner.adapter = cityArrayAdapter
                    }

                    override fun onError(anError: ANError) {
                        Log.d("SearchTutor.getCities", "onError: " + anError.errorDetail)
                    }
                })

        citySpinner.isEnabled = true

    }

    override fun onClick(v: View) {
        when (v) {
            signup_btn -> checkValidation()
            
            signup_existing_user -> 
                requireFragmentManager().transaction(now = false, allowStateLoss = false) {
                    setCustomAnimations(R.anim.left_enter, R.anim.right_enter)
                    replace(R.id.frameContainer, LoginFragment(), Utils.LOGIN_FRAGMENT)
                }
        }
    }
    

    // Check Validation Method
    private fun checkValidation() {

        // Get all edittext texts
        val fname = signup_fname.text.toString()
        val lname = signup_lname.text.toString()
        val email = signup_email.text.toString()
        val mobile = signup_mobile.text.toString()
        val city = _city.cityId
        val pswd = signup_pswd.text.toString()
        val confirmPswd = signup_confirm_pswd.text.toString()
        val mode = signup_toggle.isChecked
        
        val p = Pattern.compile(Utils.REGEX)
        val m = p.matcher(email)

        if (fname.isBlank()
                || lname.isBlank()
                || email.isBlank()
                || mobile.isBlank()
                || city <= 0
                || pswd.isBlank()
                || confirmPswd.isBlank())

            CustomToast.showToast(baseContext,
                    "All fields are required.")
        else if (!m.find())
            CustomToast.showToast(baseContext,
                    "Your Email Id is Invalid.")
        else if (confirmPswd != pswd)
            CustomToast.showToast(baseContext,
                    "Both passwords doesn't match.")
        else if (!signup_terms.isChecked)
            CustomToast.showToast(baseContext,
                    "Please accept the Terms and Conditions.")
        else {
            if (mode)
                doStudentRegister(city, fname, lname, email, mobile, pswd)
            else
                doTutorRegister(city, fname, lname, email, mobile, pswd)
        }

    }

    private fun doStudentRegister(getCity: Int, getFullName: String, getFullName2: String, getEmailId: String, getMobileNumber: String, getPassword: String) {
        AndroidNetworking.post(URLS.STUDENT_REGISTER)
                .addBodyParameter("fname", getFullName)
                .addBodyParameter("lname", getFullName2)
                .addBodyParameter("email", getEmailId)
                .addBodyParameter("mobile", getMobileNumber)
                .addBodyParameter("city", getCity.toString())
                .addBodyParameter("signup_pswd", getPassword)
                .build()
                .getAsJSONObject(object : JSONObjectRequestListener {
                    override fun onResponse(response: JSONObject) {
                        Log.d("StuSignUpFragment", "onResponse: entered")
                        try {
                            if (response.getString("status") == "accepted") {
                                Toast.makeText(context, response.getString("message"), Toast.LENGTH_SHORT).show()
                                val intent = Intent(activity, StudentDashboard::class.java)
                                Log.d("StuSignUpFragment", "onResponse: accepted")
                                startActivity(intent)
                            } else {
                                CustomToast.showToast(baseContext,
                                        response.getString("message"))
                            }
                        } catch (e: JSONException) {
                            e.printStackTrace()
                        }

                    }

                    override fun onError(anError: ANError) {
                        Log.d("StuSignUpFragment", "onError: " + anError.errorDetail)
                        Toast.makeText(context, "error", Toast.LENGTH_SHORT).show()
                    }
                })
    }

    private fun doTutorRegister(getCity: Int, getFullName: String, getFullName2: String, getEmailId: String, getMobileNumber: String, getPassword: String) {
        AndroidNetworking.post(URLS.TUTOR_REGISTER)
                .addBodyParameter("fname", getFullName)
                .addBodyParameter("lname", getFullName2)
                .addBodyParameter("email", getEmailId)
                .addBodyParameter("mobile", getMobileNumber)
                .addBodyParameter("city", getCity.toString())
                .addBodyParameter("signup_pswd", getPassword)
                .build()
                .getAsJSONObject(object : JSONObjectRequestListener {
                    override fun onResponse(response: JSONObject) {
                        Log.d("TutSignUpFragment", "onResponse: entered")
                        try {
                            if (response.getString("status") == "accepted") {
                                Toast.makeText(context, response.getString("message"), Toast.LENGTH_SHORT).show()
                                val intent = Intent(activity, SignupSegment::class.java)
                                Log.d("TutSignUpFragment", "onResponse: accepted")
                                startActivity(intent)
                            } else {
                                CustomToast.showToast(baseContext,
                                        response.getString("message"))
                            }
                        } catch (e: JSONException) {
                            e.printStackTrace()
                        }

                    }

                    override fun onError(anError: ANError) {
                        Log.d("TutSignUpFragment", "onError: " + anError.errorDetail)
                        Toast.makeText(context, "error", Toast.LENGTH_SHORT).show()
                    }
                })
    }
}
